src/graph.rs

```diff
@@ -27,27 +29,55 @@ impl GitGraph {
             repository,
             commits,
         };
-        graph.assign_branches(indices)?;
+        graph.assign_branches(indices, settings)?;
 
         Ok(graph)
     }
-    fn assign_branches(&mut self, indices: HashMap<Oid, usize>) -> Result<(), Error> {
-        let branches = self.repository.branches(None)?;
-        for branch in branches {
-            let (branch, branch_type) = branch?;
-            if branch_type == BranchType::Local {
-                let reference = branch.get();
-                if let Some(name) = reference.name() {
-                    if let Some(oid) = reference.target() {
-                        let idx = indices[&oid];
-                        self.commits[idx].branches.push(name[11..].to_string());
-                    }
+    fn assign_branches(
+        &mut self,
+        indices: HashMap<Oid, usize>,
+        settings: &Settings,
+    ) -> Result<(), Error> {
+        let branches_ordered = Self::branches_persistence_order(&self.repository, settings)?;
+
+        for branch in branches_ordered {
+            let reference = branch.get();
+            if let Some(name) = reference.name() {
+                if let Some(oid) = reference.target() {
+                    let idx = indices[&oid];
+                    self.commits[idx].branches.push(name[11..].to_string());
                 }
             }
         }
 
         Ok(())
     }
+
+    fn branches_persistence_order<'repo>(
+        repository: &'repo Repository,
+        settings: &Settings,
+    ) -> Result<Vec<Branch<'repo>>, Error> {
+        let branches = repository
+            .branches(Some(BranchType::Local))?
+            .filter_map(|bt| bt.ok().map(|bt| bt.0))
+            .map(|branch| {
+                (
+                    settings
+                        .branch_persistance
+                        .iter()
+                        .position(|b| {
+                            branch.get().name().unwrap_or("refs/heads/-")[11..].starts_with(b)
+                        })
+                        .unwrap_or(1000),
+                    branch,
+                )
+            })
+            .sorted_by_key(|(idx, _)| *idx)
```

Comment:
I think avoiding the iterator chain might be a more efficient route here. Or phrased differently, the reason that the standard library does not contain `sorted_by_key` for arbitrary iterators is the implied inefficiency as it needs to fully collect the iterator before sorting it anyway.

An alternative approach might something along the lines of
```rust
let mut branches = repository.branches(..)?.map(|b| b.map(|b| b.0)).collect::<Result<Vec<_>>()?;
  
branches.sort_by_cached_key(|b| branch_persistence.iter().position(..).unwrap_or(..));

Ok(branches)
```
where even though [`sort_by_cached_key`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.sort_by_cached_key) will also allocated auxiliary storage for the keys, it will do so in a separate allocation instead of collecting, sorting and then recollecting the elements.

Reply:
Cool, didn't know of `sort_by_cached_key`. This way, I also get rid of the `itertools` dependency. Thanks!